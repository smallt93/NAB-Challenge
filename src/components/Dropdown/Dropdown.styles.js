import styled, { css } from "styled-components";

export const DropdownWrapper = styled.ul`
  list-style-type: none;
  margin: 0;
  padding: ${({ theme }) => theme.rem(4)} ${({ theme }) => theme.rem(8)};
`;

export const DropdownItem = styled.li`
  padding: 0em .5em;
  line-height: 2rem;
  cursor: pointer;
  background-color: transparent;
  transition: all 200ms ease;
  border-radius: ${({ theme }) => theme.rem(3)};
  font-size: ${({ theme }) => theme.rem(10)};
  color: ${({ theme }) => theme.colors.textPrimary};
  ${({ theme, active }) =>
    active && css`
      background-color: ${theme.colors.greyHover};
    `
  }
`;