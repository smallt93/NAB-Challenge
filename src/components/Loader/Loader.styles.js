import styled, { css, keyframes } from 'styled-components';

const spin = keyframes`
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
`

export const LoaderWrapper = styled.div`
  border-radius: 50%;
  border: 2px solid ${({ theme }) => theme.colors.white};
  border-top: 2px solid ${({ theme }) => theme.colors.grey};
  border-bottom: 2px solid ${({ theme }) => theme.colors.grey};
  animation: ${spin} 1s linear infinite;
  position: absolute;
  right: 10px;
  z-index: 1000;
  width: 1.3rem;
  height: 1.3rem;

  ${({ theme, size }) => size === 'large' && css`
    border: 4px solid ${theme.colors.white};
    border-top: 4px solid ${theme.colors.grey};
    border-bottom: 4px solid ${theme.colors.grey};
    width: 2.5rem;
    height: 2.5rem;
    margin-bottom: 1em;
    position: unset;
  `}
`;