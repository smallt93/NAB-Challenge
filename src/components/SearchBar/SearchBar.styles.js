import styled from 'styled-components';

export const SearchBarWrapper = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: .5rem;
  padding: ${({ theme }) => theme.rem(4)} ${({ theme }) => theme.rem(8)};
  background-color: ${({ theme }) => theme.colors.white};
`;

export const SearchInput = styled.input`
  position: relative;
  border: none;
  outline: none;
  width: 100%;
  font-size:  ${({ theme }) => theme.rem(10)};
  height: ${({ theme }) => theme.rem(16)};
  color: ${({ theme }) => theme.colors.textPrimary};
  background-color: ${({ theme }) => theme.colors.white};
`;

export const LocationListWrapper = styled.div`
  position: absolute;
  z-index: 1000;
  width: 100%;
  top: 41px;
  border-radius: 1rem;
  box-shadow: 0 2px 5px 1px rgba(64,60,67,.16);
  background-color: ${({ theme }) => theme.colors.white};
`;

export const ClearSearch = styled.div`
  position: absolute;
  cursor: pointer;
  right: 18px;
  top: 3px;
  font-size: ${({ theme }) => theme.rem(12)};
  transition: all 300ms ease;
  &:hover {
    opacity: 0.7;
  }
` 