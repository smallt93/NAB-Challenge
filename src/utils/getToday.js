const weekDay = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

export const getDayOfWeek = (date) => {
  const getDay = String(date.getDate()).padStart(2, '0');
  return `${weekDay[date.getDay()]} ${getDay}`;
};