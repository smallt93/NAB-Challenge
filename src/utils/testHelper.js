import React from 'react';
import { mount, shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from 'stylesheets/defaultTheme.styles';


export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
    ),
  });
}

export function shallowWithTheme(child) {
  return shallow(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
    ),
  });
}
