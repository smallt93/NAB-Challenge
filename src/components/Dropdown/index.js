import React, { memo } from "react";
import { DropdownWrapper, DropdownItem } from './Dropdown.styles';

const Dropdown = ({
  data = [],
  activeId = 0,
  onSelectedItem = () => {},
  setActiveId = () => {},
}) => {
  const onFocusIn = (locationId) => {
    setActiveId(locationId)
  }

  const onFocusOut = () => {
    setActiveId(0)
  }

  return (
    <DropdownWrapper>
      {data.map((item, index) => (
        <DropdownItem
          key={index}
          onClick={() => onSelectedItem(item)}
          onMouseOver={() => onFocusIn(item.woeid)}
          onMouseOut={onFocusOut}
          active={activeId === item.woeid}
        >
          {item.title}
        </DropdownItem>
      ))}
    </DropdownWrapper>
  );
};

export default memo(Dropdown);
