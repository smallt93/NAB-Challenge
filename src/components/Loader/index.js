import React from 'react';
import { LoaderWrapper } from './Loader.styles';

const Loader = ({ size = '' }) => {
  return <LoaderWrapper size={size} />;
};

export default Loader;
