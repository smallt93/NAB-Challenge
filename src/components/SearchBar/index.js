import React, { memo } from 'react';
import Icon from 'components/Icon';
import Dropdown from 'components/Dropdown';
import Loader from 'components/Loader';
import iconSearch from 'assets/icons/search.svg';

import {
  SearchBarWrapper,
  LocationListWrapper,
  SearchInput,
  ClearSearch,
  Empty,
} from './SearchBar.styles';

const SearchBar = ({
  onSearching = () => {},
  onSelectedItem = () => {},
  setSearchValue = () => {},
  setActiveId = () => {},
  isSearching = false,
  isLoading = false,
  locationList = null,
  searchValue = '',
  activeId = 0,
}) => {
  return (
    <SearchBarWrapper searching={isSearching}>
      <Icon iconUrl={iconSearch} iconSize={1.5} />
      <SearchInput
        type='text'
        data-testid='search-location'
        placeholder='Search Location'
        onChange={onSearching}
        value={searchValue}
      />

      {isLoading
        ? <Loader />
        : searchValue ? (
          <ClearSearch
            className='search-reset'
            onClick={() => setSearchValue('')}
          >
            x
          </ClearSearch>
        ) : <Empty />
      }
      
      {isSearching && locationList && (
        <LocationListWrapper>
          <Dropdown
            data={locationList}
            activeId={activeId}
            onSelectedItem={onSelectedItem}
            setActiveId={setActiveId}
          />
        </LocationListWrapper>
      )}
    </SearchBarWrapper>
  );
};

export default memo(SearchBar);
