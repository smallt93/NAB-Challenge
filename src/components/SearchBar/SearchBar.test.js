import React from 'react';
import { mountWithTheme as mount } from 'utils/testHelper';
import { ClearSearch, Empty } from 'components/SearchBar/SearchBar.styles';
import SearchBar from 'components/SearchBar';
import Dropdown from 'components/Dropdown';
import Loader from 'components/Loader';

describe('render SearchBar component', () => {
  let Component;

  const mockProps = {
    onSearching: jest.fn(),
    onSelectedItem: jest.fn(),
    setSearchValue: jest.fn(),
    setActiveId: jest.fn(),
    onResetLocationList: jest.fn(),
    isSearching: false,
    isLoading: false,
    locationList: null,
    searchValue: '',
    activeId: 0,
  }
  
  beforeEach(() => {
    Component = mount(<SearchBar {...mockProps} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    expect(Component.length).toBe(1);
  });

  it('should match snapshot', () => {
    expect(Component).toMatchSnapshot();
  });

  it('should change search location value', () => {
    const SearchInput = Component.find('input[data-testid="search-location"]');
    SearchInput.simulate('change', {
      target: { value: 'Vietnam' },
    });

    expect(mockProps.onSearching).toHaveBeenCalledTimes(1);
  });

  it('should render loader component', () => {
    Component.setProps({ isLoading: true });
    const LoaderComponent = Component.find(Loader);
  
    expect(LoaderComponent.length).toBe(1);
  });

  it('should render div when search value is empty', () => {
    Component.setProps({
      isLoading: false,
      searchValue: '',
    });
    const EmptyComponent = Component.find(Empty);
  
    expect(EmptyComponent.length).toBe(1);
  });

  it('should render reset button and call setSearchValue and set empty for search value', () => {
    const expected = {
      value: '',
      trigger: false,
    }
    Component.setProps({
      isLoading: false,
      searchValue: 'Vietnam',
    });
    const ResetSearchBtn = Component.find(ClearSearch);
    expect(ResetSearchBtn.length).toBe(1);

    ResetSearchBtn.simulate('click');
    expect(mockProps.setSearchValue).toHaveBeenCalledTimes(1);
    expect(mockProps.setSearchValue).toHaveBeenCalledWith(expected);
    expect(mockProps.onResetLocationList).toHaveBeenCalledTimes(1);
  });

  it('should render null when search value and locationList are empty', () => {
    Component.setProps({
      isSearching: false,
      locationList: null,
    });
    const DropdownComponent = Component.find(Dropdown);
    expect(DropdownComponent.length).toBe(0);
  });

  it('should render dropdown location list', () => {
    Component.setProps({
      isSearching: true,
      locationList: [{}, {}, {}],
    });
    const DropdownComponent = Component.find(Dropdown);
    expect(DropdownComponent.length).toBe(1);
  });
});
