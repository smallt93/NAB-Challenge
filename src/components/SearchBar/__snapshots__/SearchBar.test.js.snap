// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SearchBar component should match snapshot 1`] = `ReactWrapper {}`;
